### ISSUE TITLE
How to reproduce the 'oracle' results?

### ISSUE THREAD

**Issue by 1160300911** [tzshi/squall#3]
Hi, first of all, thanks for sharing the source code and data.

The problem is when I want to use the 'oracle' setting with BERT encoder, I found the '--gold-attn' argument made slight change on the performance, not as large as mentioned in the paper. So I'm wondering if I made some mistakes on parameter setting? Besides, I found that at the encoding part of the model (maybe '_bert_features' function?), there is no code corresponding to the '--gold-attn' argument.

Looking forward to your reply.

**Comment by tzshi** [comment]
Hi,

Thank you for your interest in our work!

A closer check suggested that this released code base does not contain all the code for running the oracle experiments. We will work on updating this code base.

I am including below the code snippets we used for the encoder-attention oracle experiments in the paper, in case you want it more urgently than we can fully re-test and update the code. (Chen and I currently occupied with other projects in the next few days; let us know if you run into any issues.)

Replace Lines 396-399 with the following (We only did LSTM version in the paper, but I think BERT features should be handled similarly, at lines 274-275):
```
        if self.args.gold_attn_enc:
            wcontext_list = list()
            ccontext_list = list()

            for i, instance in enumerate(batch['instances']):
                w_att_c_matrix, c_att_w_matrix = attention_matrix_col(instance)
                wcontext_i, alpha_w = self.w_attn_c(wvecs[i].unsqueeze(0), cvecs[i].unsqueeze(0), c_mask[i].unsqueeze(0), gold_attn=w_att_c_matrix, device=self.device)
                wcontext_list.append(wcontext_i.squeeze(0))
        
                ccontext_i, alpha_c = self.c_attn_w(cvecs[i].unsqueeze(0), wvecs[i].unsqueeze(0), w_mask[i].unsqueeze(0), gold_attn=c_att_w_matrix, device=self.device)
                ccontext_list.append(ccontext_i.squeeze(0))
            wcontext = torch.stack(wcontext_list)
            ccontext = torch.stack(ccontext_list)
        else:
            wcontext, alpha_w = self.w_attn_c(wvecs, cvecs, c_mask)
            ccontext, alpha_c = self.c_attn_w(cvecs, wvecs, w_mask)
```

Replace `forward` function in the `MatchAttn` class (lines 802 forward) with:
```
    def forward(self, x, y, y_mask, is_score=False, gold_attn=None, device=None):
        # Project vectors
        if self.linear_x:
            x_proj = self.linear_x(x.view(-1, x.size(2))).view(x.shape[0], x.shape[1], -1)
            x_proj = F.relu(x_proj)
            y_proj = self.linear_y(y.view(-1, y.size(2))).view(y.shape[0], y.shape[1], -1)
            y_proj = F.relu(y_proj)
        else:
            x_proj = x
            y_proj = y

        x_proj = self.w(x_proj)

        # Compute scores
        scores = x_proj.bmm(y_proj.transpose(2, 1))
        y_mask = y_mask.unsqueeze(1).expand(scores.size())
        if gold_attn is not None:
            new_score_list = list()
            for idx, ii in enumerate(gold_attn.tolist()):
                for tt in range(len(ii), scores.size(-1)):
                    ii.append(0.0)
                if np.sum(ii) == 0:
                    num_cols = 0
                    for mask_item in y_mask[0][idx]:
                        if mask_item == 0:
                            num_cols += 1
                    new_score_list.append(torch.tensor(len(ii) * [1.0 / num_cols]))
                else:
                    new_score_list.append(torch.tensor(ii))
            for tt in range(len(gold_attn.tolist()), scores.size(1)):
                new_score_list.append(torch.tensor(scores.size(-1) * [0.0]))

            scores = torch.stack(new_score_list).unsqueeze(0).to(device)
        # Mask padding
        scores.data.masked_fill_(y_mask.bool().data, -float('inf'))
        # Normalize with softmax
        if is_score:
            return scores
        alpha_flat = F.softmax(scores.view(-1, y.size(1)), dim=-1)
        alpha = alpha_flat.view(-1, x.size(1), y.size(1))

        # Take weighted average
        matched_seq = alpha.bmm(y)

        return matched_seq, alpha
```

**Comment by 1160300911** [comment]
Hi, thanks for your apply!

Under your suggestions above, I changed the source code, and got the following results on the dev-0 set:

methods | ACC(LF) | ACC(EXE)
---- | ---  | ---
ALIGN w/BERT | 46.52715939447907 | 65.09349955476402
\+ Encoder oracle |  52.09260908281389 | 68.78895814781835
\+ Decoder oracle | 47.10596616206589 | 60.77471059661621
\+ Both oracle | 52.67141585040071 | 65.27159394479074

It seems that the 'oracle' signal has worked at the encoding part, although it is not as obvious as without using BERT. Thanks for your suggestions! By the way, there are some bugs in the `forward` function above. Here is the fixed version (not sure if it is completely correct):

```
    def forward(self, x, y, y_mask, is_score=False, gold_attn=None, device=None):
        # Project vectors
        if self.linear_x:
            x_proj = self.linear_x(x.view(-1, x.size(2))).view(x.shape[0], x.shape[1], -1)
            x_proj = F.relu(x_proj)
            y_proj = self.linear_y(y.view(-1, y.size(2))).view(y.shape[0], y.shape[1], -1)
            y_proj = F.relu(y_proj)
        else:
            x_proj = x
            y_proj = y

        x_proj = self.w(x_proj)

        # Compute scores
        scores = x_proj.bmm(y_proj.transpose(2, 1))
        y_mask = y_mask.unsqueeze(1).expand(scores.size())

        # Mask padding
        scores.data.masked_fill_(y_mask.bool().data, -float('inf'))
        # Normalize with softmax
        if is_score:
            return scores
        alpha_flat = F.softmax(scores.view(-1, y.size(1)), dim=-1)
        alpha = alpha_flat.view(-1, x.size(1), y.size(1))

        # MODIFIED (gold_attn should after the 'softmax' function)
        if gold_attn is not None:
            new_score_list = list()
            for idx, ii in enumerate(gold_attn.tolist()):
                for tt in range(len(ii), scores.size(-1)):
                    ii.append(0.0)
                if np.sum(ii) == 0:
                    num_cols = 0
                    for mask_item in y_mask[0][idx]:
                        if mask_item == 0:
                            num_cols += 1
                    new_score_list.append(torch.tensor(len(ii) * [1.0 / num_cols]))
                else:
                    new_score_list.append(torch.tensor(ii))
            for tt in range(len(gold_attn.tolist()), scores.size(1)):
                new_score_list.append(torch.tensor(scores.size(-1) * [1.0 / scores.size(-1)]))

            alpha = torch.stack(new_score_list).unsqueeze(0).to(device)

        # Take weighted average
        matched_seq = alpha.bmm(y)

        return matched_seq, alpha
```

However, the performance of the 'decoder oracle' setting still looks abnormal. Specifically, I simulated this setting by using the `--gold-attn` argument. So I'm wondering if I made some mistakes during the process? Could you please give me some advice about that? Thanks!

**Comment by tzshi** [comment]
Thank you for pointing out the ordering issue. Yes, the `gold_attn` portion should come after softmax. And glad to see that the encoder oracle attention seems to be working as expected.

For the decoder part, I don't see any initial suspect at the moment. Chen and I will get back to this issue in a few days.

(While in the paper, we only tested the LSTM encoder, I don't think the oracle decoder attention should break using the BERT encoder. 

One variation we did try in our code was to have these two lines in `utils.py` after line 132
```
    else:
        att[i] = 1./len(att[i])
```

But I am not sure this would create any huge difference.

Anyway, we will try to get back to you in a few days, after the EMNLP submission deadline.)

Thank you for your patience!

**Comment by henryzhao5852** [comment]
Hi,

I have tried running decoder oracle with BERT (python main.py --bert --gold-attn), the EM score is 52.49 after 5 iterations (it could increase in later iters). So I think the code is probably correct. A tip for debugging is to print the ``decode_w_attn.attn'', check whether it equals to the attention matrix.

If you still have problems on using oracle for decoding, could you share the log file so it would be easier to do debugging (i.e., loss/accu at each iteration)? 

Best,
Chen

**Comment by 1160300911** [comment]
Hi, sorry for the late reply! @tzshi @henryzhao5852 

After adding the two lines after line 132, the performance of the 'decoder oracle' setting has changed from 47.10596616206589 to 50.84594835262689 on ACC(LF) without retraining. And I get 56.94568121104185 ACC(LF) score when using the 'both oracle' setting with BERT encoder. The performance has indeed improved, but it is stll lower than the result in the paper without BERT.

Besides, I also have tried running decoder oracle using the command mentioned above `python main.py --bert --gold-attn`, and the ACC(LF) score is 53.38379341050757 after 19 epoches. So I'm wondering whether the results in the paper are obtained by retraining the model? 

By reading the release code, I found that if the model is retrained using the `--gold-attn` argument, it means that the model will use the gold attention matrix during the forward process, not only for calculating the training loss. Is this the keypoint to reproduce the performance in the paper? 

Looking forward to your reply. Thanks!

**Comment by henryzhao5852** [comment]
Hi,

Yes, oracle experiments assume the annotated (oracle) alignments are given both on training and evaluation. So we train the model from scratch, using oracle alignment (so no loss for decoder attention etc), and alignments are directly used for inference. 

Best,
Chen
